lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -168,6 +208,75 @@ impl HyperV {
             }
         }
     }
+
+    fn handle_rdmsr_time_ref_count(&self) -> RdmsrOutcome {
+        if !self.features.reference_tsc {
+            return RdmsrOutcome::GpException;
+        }
+
+        let time_data = vmm::time::export_time_data(self.vmm_hdl())
+            .expect("VMM time data can always be exported");
+
+        // `hrtime` is the current host high-resolution timer value (in
+        // nanoseconds), and `boot_hrtime` is the high-resolution timer value at
+        // which the VM was started, so it suffices to subtract the latter from
+        // the former and divide by 100 to get 100-nanosecond units.
+        //
+        // If this VM is migrated or otherwise saved/restored, the migration
+        // protocol is expected to set `boot_hrtime` for the restored VM to
+        // account for any hrtime differences between hosts and any time where
+        // the VM was paused.
```

Comment:
I don't think it's true that it takes into account pause time explicitly; the adjustment on the propolis side for live migration is to reconcile the boot_hrtime on a new host (relative to a new host hrtime), and on the kernel side, the TSC is adjusted, but not the boot_hrtime. The time of migration is accounted for in these calculations, but not so explicity.

https://github.com/oxidecomputer/illumos-gate/blob/7633a05bff8c639f2df722d1fba7b889b2763d3d/usr/src/uts/intel/io/vmm/vmm.c#L4999-L5031

Reply:
I knocked this language out in dce0275. I had in mind a case where one has a running VM and chooses to "suspend" it so that its state is saved and it no longer perceives the passage of time. In a case like that I think you'd need to adjust `boot_hrtime` on "resume" to give the illusion that no time passed while the VM was paused. But that's not a use case we have today, so I've simplified this comment just to talk about migration.